bigtable_input.py

```diff
@@ -97,76 +73,259 @@ def make_single_array(ds, batch_size=8*1024):
+        utils.dbg('Row range: %s - %s; total moves: %d; probability %.3f' % (
+            start_row, end_row, total_moves, probability))
+        shards = 8
+        ds = self.tf_table.parallel_scan_range(start_row, end_row,
+                                               probability=probability,
+                                               num_parallel_scans=shards,
+                                               columns=[(column_family, column)])
+        if shuffle:
+            rds = tf.data.Dataset.from_tensor_slices(
+                tf.random_shuffle(tf.range(0, shards, dtype=tf.int64)))
+            ds = rds.apply(
+                tf.contrib.data.parallel_interleave(
+                    lambda x: ds.shard(shards, x),
+                    cycle_length=shards, block_length=1024))
+            ds = ds.shuffle(shards * 1024 * 2)
+        ds = ds.take(moves)
+        return ds
+
+    def _write_move_counts(self, sess, h):
+        """Add move counts from the given histogram to the table.
+
+        Used to update the move counts in an existing table.  Should
+        not be needed except for backfill or repair.
+
+        Args:
+          sess:  TF session to use for doing a Bigtable write.
+          tf_table:  TF Cloud Bigtable to use for writing.
+          h:  a dictionary keyed by game row prefix ("g_0023561") whose values
+             are the move counts for each game.
+        """
+        def gen():
+            for k, v in h.items():
+                # The keys in the histogram may be of type 'bytes'
+                k = str(k, 'utf-8')
+                vs = str(v)
+                yield (k.replace('g_', 'ct_') + '_%d' % v, vs)
+                yield (k + '_m_000', vs)
+        mc = tf.data.Dataset.from_generator(gen, (tf.string, tf.string))
+        wr_op = self.tf_table.write(mc,
+                                    column_families=[METADATA],
+                                    columns=[MOVE_COUNT])
+        sess.run(wr_op)
+
+    def update_move_counts(self, start_game, end_game, interval=1000):
+        """Used to update the move_count cell for older games.
+
+        Should not be needed except for backfill or repair.
+
+        move_count cells will be updated in both g_<game_id>_m_000 rows
+        and ct_<game_id>_<move_count> rows.
+        """
+        for g in range(start_game, end_game, interval):
+            with tf.Session() as sess:
+                start_row = ROW_PREFIX.format(g)
+                end_row = ROW_PREFIX.format(g + interval)
+                print('Range:', start_row, end_row)
+                start_time = time.time()
+                ds = self.tf_table.keys_by_range_dataset(start_row, end_row)
+                h = _histogram_move_keys_by_game(sess, ds)
+                self._write_move_counts(sess, h)
+                end_time = time.time()
+                elapsed = end_time - start_time
+                print('  games/sec:', len(h)/elapsed)
+
+
+# TODO(president.jackson): document these
```

Comment:
document what parts?  for who?  PROJECT is the cloud project id...

Reply:
I believe this is a comment that Seth added when the first draft of this went in, and the refactor has made it visible again. I can write a quick blurb for each; where do we like this documentation?

Reply:
ok so my question when i read this request for docs is is "who is looking for this and when do they see it -- i.e., does someone running this without them set get an error?  what is the person reading it trying to accomplish?"  etc.

it looks like this will give a KeyError if any of these aren't defined, so probably a little blurb here about the CBT_ variables is fine.

Reply:
Done!